firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -107,24 +126,180 @@ private void applyLocalMutationsToDocuments(
    */
   ImmutableSortedMap<DocumentKey, Document> getDocuments(Iterable<DocumentKey> keys) {
     Map<DocumentKey, MutableDocument> docs = remoteDocumentCache.getAll(keys);
-    return getLocalViewOfDocuments(docs);
+    Map<DocumentKey, MutableDocument> local_docs = localDocumentCache.getAll(keys);
+    ImmutableSortedMap<DocumentKey, Document> results = emptyDocumentMap();
+
+    for (Map.Entry<DocumentKey, MutableDocument> entry : docs.entrySet()) {
+      if (local_docs.containsKey(entry.getKey())) {
+        MutableDocument localDoc = local_docs.get(entry.getKey());
+        SnapshotVersion version = entry.getValue().getVersion();
+        if (localDoc.getVersion().compareTo(version) > 0) {
+          version = localDoc.getVersion();
+        }
+        results = results.insert(entry.getKey(), localDoc.setVersion(version));
+      } else {
+        results = results.insert(entry.getKey(), entry.getValue());
+      }
+    }
+    return results;
+  }
+
+  public void saveLocalDocuments(ImmutableSortedMap<DocumentKey, Document> documents) {
+    for (Map.Entry<DocumentKey, Document> entry : documents) {
+      // TODO(Overlay): This cast is ugly.
+      localDocumentCache.add((MutableDocument) entry.getValue());
+    }
+  }
+
+  public void repopulateCache(Set<DocumentKey> keys) {
+    Map<DocumentKey, MutableDocument> existingDocs = remoteDocumentCache.getAll(keys);
+
+    List<MutationBatch> batches = mutationQueue.getAllMutationBatchesAffectingDocumentKeys(keys);
+    // TODO(Overlay): This is running for all changed docs X all batches that might affect..pretty
```

Comment:
As suggested yesterday: 
- We only need to recompute the cache for documents with transforms.
- We might be able to remove some recomputations if we batch document updates (in the hope that there are duplicate updates).

Reply:
Right.